### PR TITLE
Firming up client stream closure scenarios.

### LINKED ISSUE #536: Investigate Netty stream closure handling logic
In the Netty handlers, we seem to have various ways that we can detect stream and handle stream closure. We should try to simplify this logic so that it's less error prone.

### PR DESCRIPTION
**Pull request by nmittler**
- AbstractClientStream
-- Properly set the inboundPhase after receiving trailers. This is needed to allow the generic onStreamClosed() handler properly report the status for the closed stream.

- BufferingHttp2ConnectionEncoder
-- Moved exceptions to inner classes.
-- All new headers fail with new exception type after encoder has been closed.

- NettyClientHandler
-- Closing the buffering encoder when the channel goes inactive in order to explicitly clear out any pending streams.
-- Handling onStreamClosed() event to ensure that we always report status for closed streams.
-- Added handling for stream exhaustion.  Immediately fail new streams with a reasonable error.
-- Using a second promise in createStream() to ensure that the completion handler in this method runs before the handler in the transport.

Fixes #536

### CODE REVIEW

**Comment by nmittler** [comment]
@ejona86 PTAL

**Comment by nmittler** [comment]
@ejona86 ping...

File: core/src/main/java/io/grpc/transport/AbstractClientStream.java [hunk 1]
```diff
@@ -173,6 +173,7 @@ protected void inboundTrailersReceived(Metadata.Trailers trailers, Status status
     this.status = status;
     this.trailers = trailers;
     deframe(ReadableBuffers.empty(), true);
+    inboundPhase(Phase.STATUS);
```
Comment (ejona86) [inline core/src/main/java/io/grpc/transport/AbstractClientStream.java]: This should be set as early as possible (so immediately after the if). Otherwise any reentrant calls will see the wrong phase. That is already being for the other inboundPhase()s in this class.

File: netty/src/main/java/io/grpc/transport/netty/BufferingHttp2ConnectionEncoder.java [hunk 2]
```diff
@@ -67,6 +69,44 @@
  * in replacement for {@link io.netty.handler.codec.http2.DefaultHttp2ConnectionEncoder}.
  */
 class BufferingHttp2ConnectionEncoder extends DecoratingHttp2ConnectionEncoder {
+
+  /**
+   * Thrown if buffered streams are terminated due to this encoder being closed.
+   */
+  public static final class ChannelClosedException extends Http2Exception {
+    public ChannelClosedException() {
+      super(Http2Error.STREAM_CLOSED, "Connection closed");
```
Comment (ejona86) [inline netty/src/main/java/io/grpc/transport/netty/BufferingHttp2ConnectionEncoder.java]: STREAM_CLOSED is for more of a protocol error. It isn't applicable here.

REFUSED_STREAM seems more appropriate.

File: netty/src/main/java/io/grpc/transport/netty/BufferingHttp2ConnectionEncoder.java [hunk 3]
```diff
@@ -67,6 +69,44 @@
  * in replacement for {@link io.netty.handler.codec.http2.DefaultHttp2ConnectionEncoder}.
  */
 class BufferingHttp2ConnectionEncoder extends DecoratingHttp2ConnectionEncoder {
+
+  /**
+   * Thrown if buffered streams are terminated due to this encoder being closed.
+   */
+  public static final class ChannelClosedException extends Http2Exception {
```
Comment (ejona86) [inline netty/src/main/java/io/grpc/transport/netty/BufferingHttp2ConnectionEncoder.java]: Why do these extend Http2Exception? I'm not sure it works well, given that normally we use Http2Exceptions for determining what to _send_ but here we are using them for informing of receipt. That makes the Http2Errors hard to specify, since we don't want to communicate an error to the remote.

File: netty/src/main/java/io/grpc/transport/netty/BufferingHttp2ConnectionEncoder.java [hunk 4]
```diff
@@ -67,6 +69,44 @@
...
+    }
+  }
+
+  /**
+   * Thrown if buffered streams are terminated due to receipt of a {@code GOAWAY}.
+   */
+  public static final class GoAwayException extends Http2Exception {
+    private final int lastStreamId;
+    private final long errorCode;
+    private final ByteBuf debugData;
```
Comment (ejona86) [inline netty/src/main/java/io/grpc/transport/netty/BufferingHttp2ConnectionEncoder.java]: Nothing for this PR, but we should realize that this ByteBuf is impossible to release correctly.

File: netty/src/main/java/io/grpc/transport/netty/BufferingHttp2ConnectionEncoder.java [hunk 5]
```diff
@@ -195,7 +241,19 @@ public ChannelFuture writeSettingsAck(ChannelHandlerContext ctx, ChannelPromise
...
   public void close() {
     super.close();
-    cancelPendingStreams();
+
+    if (!closed) {
+      closed = true;
+
+      // Fail all buffered streams.
+      ChannelClosedException e = new ChannelClosedException();
+      PendingStream[] streams = pendingStreams.values().toArray(
```
Comment (ejona86) [inline netty/src/main/java/io/grpc/transport/netty/BufferingHttp2ConnectionEncoder.java]: Am I correct in reading the code, that this copy is unnecessary and this loop operates basically the same as the removed `cancelPendingStreams()`?

**Comment by ejona86** [comment]
Didn't finish review. I still have to look closely at `NettyClientHandler`.

File: core/src/main/java/io/grpc/transport/AbstractClientStream.java (hunk 1 above)
Comment (nmittler) [inline core/src/main/java/io/grpc/transport/AbstractClientStream.java]: done.

File: netty/src/main/java/io/grpc/transport/netty/BufferingHttp2ConnectionEncoder.java (hunk 3 above)
Comment (nmittler) [inline netty/src/main/java/io/grpc/transport/netty/BufferingHttp2ConnectionEncoder.java]: This is primarily due to the fact that we're going to eventually migrate to the Netty version of this class which throws Http2Exceptions.

File: netty/src/main/java/io/grpc/transport/netty/BufferingHttp2ConnectionEncoder.java (hunk 2 above)
Comment (nmittler) [inline netty/src/main/java/io/grpc/transport/netty/BufferingHttp2ConnectionEncoder.java]: Done.

File: netty/src/main/java/io/grpc/transport/netty/BufferingHttp2ConnectionEncoder.java (hunk 4 above)
Comment (nmittler) [inline netty/src/main/java/io/grpc/transport/netty/BufferingHttp2ConnectionEncoder.java]: Yeah, probably should just convert to a UTF_8 string before creating the exception. WDYT?

File: netty/src/main/java/io/grpc/transport/netty/BufferingHttp2ConnectionEncoder.java (hunk 5 above)
Comment (nmittler) [inline netty/src/main/java/io/grpc/transport/netty/BufferingHttp2ConnectionEncoder.java]: Reworked this so that we use foreach loop and clear a the end.

**Comment by nmittler** [comment]
@ejona86 I've addressed the comments so far.

File: netty/src/main/java/io/grpc/transport/netty/BufferingHttp2ConnectionEncoder.java (hunk 4 above)
Comment (ejona86) [inline netty/src/main/java/io/grpc/transport/netty/BufferingHttp2ConnectionEncoder.java]: That could be reasonable. It doesn't _have_ to be UTF-8 though. Maybe ByteString?

**Comment by ejona86** [comment]
@nmittler, FYI, you didn't push the new version. I still have plenty to review on previous version though :)

**Comment by nmittler** [comment]
@ejona86 oops :) ... pushed the changes.

File: netty/src/main/java/io/grpc/transport/netty/BufferingHttp2ConnectionEncoder.java [hunk 6]
```diff
@@ -67,6 +69,44 @@
...
+  /**
+   * Thrown if buffered streams are terminated due to receipt of a {@code GOAWAY}.
+   */
+  public static final class GoAwayException extends Http2Exception {
+    private final int lastStreamId;
+    private final long errorCode;
+    private final ByteBuf debugData;
+
+    public GoAwayException(int lastStreamId, long errorCode, ByteBuf debugData) {
+      super(Http2Error.STREAM_CLOSED);
```
Comment (ejona86) [inline netty/src/main/java/io/grpc/transport/netty/BufferingHttp2ConnectionEncoder.java]: REFUSED_STREAM here, too.

File: netty/src/main/java/io/grpc/transport/netty/NettyClientHandler.java [hunk 7]
```diff
@@ -111,9 +116,22 @@ public NettyClientHandler(BufferingHttp2ConnectionEncoder encoder, Http2Connecti
...
     connection.addListener(new Http2ConnectionAdapter() {
       @Override
+      public void onStreamClosed(Http2Stream stream) {
+        // Report the closure status for the stream.
+        reportStatus(clientStream(stream), null, false);
+      }
+
+      @Override
       public void onGoAwayReceived(int lastStreamId, long errorCode, ByteBuf debugData) {
         goAwayStatus(statusFromGoAway(errorCode, debugData));
```
Comment (ejona86) [inline netty/src/main/java/io/grpc/transport/netty/NettyClientHandler.java]: Are we sure this is called before `onStreamClosed()` for any streams that will be closed due to the GOAWAY?

File: netty/src/main/java/io/grpc/transport/netty/NettyClientHandler.java [hunk 8]
```diff
@@ -203,13 +221,35 @@ private void onDataRead(int streamId, ByteBuf data, boolean endOfStream) throws
...
    * Handler for an inbound HTTP/2 RST_STREAM frame, terminating a stream.
    */
   private void onRstStreamRead(int streamId)
       throws Http2Exception {
     // TODO(nmittler): do something with errorCode?
-    NettyClientStream stream = clientStream(requireHttp2Stream(streamId));
-    stream.transportReportStatus(Status.UNKNOWN, false, new Metadata.Trailers());
+    // Only need to report status, the base handler will close the stream after returning
+    // from this method.
+    reportStatus(clientStream(requireHttp2Stream(streamId)), null, false);
```
Comment (ejona86) [inline netty/src/main/java/io/grpc/transport/netty/NettyClientHandler.java]: Why use reportStatus? It seems we should use the actual errorCode that came with the RST_STREAM. This error is likely unrelated to the GOAWAY.

File: netty/src/main/java/io/grpc/transport/netty/NettyClientHandler.java [hunk 9]
```diff
@@ -225,18 +265,17 @@ public void close(ChannelHandlerContext ctx, ChannelPromise promise) throws Exce
...
-      connection().forEachActiveStream(new Http2StreamVisitor() {
-        @Override
-        public boolean visit(Http2Stream stream) throws Http2Exception {
-          clientStream(stream).transportReportStatus(goAwayStatus, false, new Metadata.Trailers());
-          return true;
-        }
-      });
+
+      // Close the buffered encoder and cancel any pending streams.
+      encoder().close();
```
Comment (ejona86) [inline netty/src/main/java/io/grpc/transport/netty/NettyClientHandler.java]: Why do we need this? It seems like Http2ConnectionHandler does this.

File: netty/src/main/java/io/grpc/transport/netty/NettyClientHandler.java [hunk 10]
```diff
@@ -260,11 +299,11 @@ protected void onStreamError(ChannelHandlerContext ctx, Throwable cause,
     // Close the stream with a status that contains the cause.
     Http2Stream stream = connection().stream(http2Ex.streamId());
     if (stream != null) {
-      clientStream(stream).transportReportStatus(Status.fromThrowable(cause), false,
-          new Metadata.Trailers());
+      // Report the status for this stream.
+      reportStatus(clientStream(stream), Status.fromThrowable(cause), false);
```
Comment (ejona86) [inline netty/src/main/java/io/grpc/transport/netty/NettyClientHandler.java]: Why this change? It seems most likely this error is unrelated to GOAWAY.

File: netty/src/main/java/io/grpc/transport/netty/NettyClientHandler.java [hunk 11]
```diff
@@ -280,32 +319,61 @@ protected boolean isGracefulShutdownComplete() {
...
+      streamId = getAndIncrementNextStreamId();
+    } catch (StatusException e) {
+      // Stream IDs have been exhausted for this connection. Fail the promise immediately.
+      promise.setFailure(e);
+      return;
+    }
+
     final NettyClientStream stream = command.stream();
     final Http2Headers headers = command.headers();
-    // TODO: Send GO_AWAY if streamId overflows
```
Comment (ejona86) [inline netty/src/main/java/io/grpc/transport/netty/NettyClientHandler.java]: Why remove this TODO? We still aren't stopping the transport if streamId overflowed.

File: netty/src/main/java/io/grpc/transport/netty/NettyClientHandler.java [hunk 12]
```diff
@@ -318,7 +386,8 @@ public void operationComplete(ChannelFuture future) throws Exception {
   private void cancelStream(ChannelHandlerContext ctx, CancelStreamCommand cmd,
       ChannelPromise promise) {
     NettyClientStream stream = cmd.stream();
-    stream.transportReportStatus(Status.CANCELLED, true, new Metadata.Trailers());
+    reportStatus(stream, Status.CANCELLED, true);
```
Comment (ejona86) [inline netty/src/main/java/io/grpc/transport/netty/NettyClientHandler.java]: Why call reportStatus? Again, it doesn't seem GOAWAY's status is appropriate.

File: netty/src/main/java/io/grpc/transport/netty/NettyClientHandler.java [hunk 13]
```diff
@@ -433,11 +481,12 @@ private Throwable getPingFailure() {
...
   private Status statusFromGoAway(long errorCode, ByteBuf debugData) {
     Status status = HttpUtil.Http2Error.statusForCode((int) errorCode);
-    if (debugData.isReadable()) {
-      // If a debug message was provided, use it.
-      String msg = debugData.toString(UTF_8);
+    String msg = debugData.toString(UTF_8);
+    if (!msg.isEmpty()) {
       status = status.augmentDescription(msg);
     }
+    logger.log(Level.FINE, "Received GOAWAY from server. Code=%d, Message=%s",
```
Comment (ejona86) [inline netty/src/main/java/io/grpc/transport/netty/NettyClientHandler.java]: Logger isn't printf-style, it is MessageFormat style. Use {0} and {1}

File: netty/src/main/java/io/grpc/transport/netty/NettyClientHandler.java [hunk 14]
```diff
@@ -448,7 +497,18 @@ private NettyClientStream clientStream(Http2Stream stream) {
...
+  /**
+   * Gets the next valid stream ID for a new stream.
+   *
+   * @throws StatusException thrown if the stream IDs for this connection have been exhausted.
+   */
+  private int getAndIncrementNextStreamId() throws StatusException {
+    if (nextStreamId < 0) {
+      // Set the GOAWAY status, but don't overwrite.
+      goAwayStatus(EXHAUSTED_STREAMS_STATUS);
+      throw goAwayStatus.asException();
```
Comment (ejona86) [inline netty/src/main/java/io/grpc/transport/netty/NettyClientHandler.java]: Should this be EXHAUSTED_STREAMS_STATUS instead of goAwayStatus?

File: netty/src/test/java/io/grpc/transport/netty/NettyHandlerTestBase.java [hunk 15]
```diff
@@ -129,7 +131,7 @@ protected final ChannelHandlerContext newContext() {
   }
 
   protected final ChannelPromise newPromise() {
-    return Mockito.mock(ChannelPromise.class);
+    return new DefaultChannelPromise(channel, ImmediateEventExecutor.INSTANCE);
```
Comment (ejona86) [inline netty/src/test/java/io/grpc/transport/netty/NettyHandlerTestBase.java]: Excellent.

File: netty/src/main/java/io/grpc/transport/netty/BufferingHttp2ConnectionEncoder.java (hunk 4 above)
Comment (nmittler) [inline netty/src/main/java/io/grpc/transport/netty/BufferingHttp2ConnectionEncoder.java]: done.

File: netty/src/main/java/io/grpc/transport/netty/BufferingHttp2ConnectionEncoder.java (hunk 6 above)
Comment (nmittler) [inline netty/src/main/java/io/grpc/transport/netty/BufferingHttp2ConnectionEncoder.java]: done.

File: netty/src/main/java/io/grpc/transport/netty/NettyClientHandler.java (hunk 7 above)
Comment (nmittler) [inline netty/src/main/java/io/grpc/transport/netty/NettyClientHandler.java]: Yes this callback occurs before the streams are closed: https://github.com/netty/netty/blob/ed4a89082bb29b9e7d869c5d25d6b9ea8fc9d25b/codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionDecoder.java#L172

File: netty/src/main/java/io/grpc/transport/netty/NettyClientHandler.java (hunk 8 above)
Comment (nmittler) [inline netty/src/main/java/io/grpc/transport/netty/NettyClientHandler.java]: We were just using UKNOWN before ... are you asking to maintain that or to use a mapping from the HTTP/2 error code?

File: netty/src/main/java/io/grpc/transport/netty/NettyClientHandler.java (hunk 9 above)
Comment (nmittler) [inline netty/src/main/java/io/grpc/transport/netty/NettyClientHandler.java]: This is the main thrust of this PR. We're clearing out the buffering encoder's cached streams before we delegate to the base class to close any active streams.  Otherwise the buffering encoder has to rely on repeated callbacks from stream closures to clear itself out one stream at a time.

File: netty/src/main/java/io/grpc/transport/netty/NettyClientHandler.java (hunk 11 above)
Comment (nmittler) [inline netty/src/main/java/io/grpc/transport/netty/NettyClientHandler.java]: Modified the logic at the top of createStream to begin graceful shutdown of the channel.

File: netty/src/main/java/io/grpc/transport/netty/NettyClientHandler.java (hunk 8 above)
Comment (ejona86) [inline netty/src/main/java/io/grpc/transport/netty/NettyClientHandler.java]: We are mainly going to be using UNKNOWN now too, except if there was an (unrelated) goaway, then we start giving some random status.

If we are going to change it, it seems we should base it on the received HTTP/2 error code.

File: netty/src/main/java/io/grpc/transport/netty/NettyClientHandler.java (hunk 9 above)
Comment (ejona86) [inline netty/src/main/java/io/grpc/transport/netty/NettyClientHandler.java]: Okay, I see it now.

**Comment by nmittler** [comment]
@ejona86 I think I'm going to break this PR into parts that are less controversial.  Stay tuned.

**Comment by nmittler** [comment]
Split out the non-controversial bits into #585 and #590

The rest I think we should abandon for now. Closing this issue.
